src/CoreWCF.Kafka/src/CoreWCF/Channels/KafkaTransportPump.cs

```diff
@@ -0,0 +1,246 @@
+        {
+            Producer.Flush(CancellationToken.None);
+        }
+        Consumer.Close();
+    }
+
+    private static (bool? EnableAutoCommit, bool? EnableAutoOffsetStore) GetCommitStrategyConfigValues(ConsumerConfig consumerConfig, KafkaDeliverySemantics kafkaDeliverySemantics) =>
+        (kafkaDeliverySemantics, consumerConfig) switch
+        {
+            // KafkaBinding
+            (KafkaDeliverySemantics.AtMostOnce, { EnableAutoCommit: null, EnableAutoOffsetStore: null }) => s_atMostOnceConfigValues,
+            (KafkaDeliverySemantics.AtLeastOnce, { EnableAutoCommit: null, EnableAutoOffsetStore: null } ) => s_atLeastOnceBatchCommitConfigValues,
+            // CustomBinding
+            (KafkaDeliverySemantics.AtMostOnce, { EnableAutoCommit: false, EnableAutoOffsetStore: null }) => s_atMostOnceConfigValues,
+            (KafkaDeliverySemantics.AtLeastOnce, { EnableAutoCommit: true, EnableAutoOffsetStore: false } ) => s_atLeastOnceBatchCommitConfigValues,
+            (KafkaDeliverySemantics.AtLeastOnce, { EnableAutoCommit: false, EnableAutoOffsetStore: null }) => s_atLeastOncePerMessageCommitConfigValues,
+            _ => throw new NotSupportedException(string.Format(SR.InvalidKafkaConfiguration, kafkaDeliverySemantics, consumerConfig.EnableAutoCommit, consumerConfig.EnableAutoOffsetStore))
+        };
+
+    private void OnLog(IConsumer<Null, byte[]> consumer, LogMessage logMessage)
+    {
+        const string format = "{0}:{1}";
+        switch (logMessage.Level)
+        {
+            case SyslogLevel.Debug:
+                _logger.LogDebug(format, logMessage.Name, logMessage.Message);
+                break;
+            case SyslogLevel.Notice:
+            case SyslogLevel.Info:
+                _logger.LogInformation(format, logMessage.Name, logMessage.Message);
+                break;
+            case SyslogLevel.Warning:
+                _logger.LogWarning(format, logMessage.Name, logMessage.Message);
+                break;
+            case SyslogLevel.Error:
+                _logger.LogError(format, logMessage.Name, logMessage.Message);
+                break;
+            case SyslogLevel.Alert:
+            case SyslogLevel.Critical:
+            case SyslogLevel.Emergency:
+                _logger.LogCritical(format, logMessage.Name, logMessage.Message);
+                break;
+            default:
+                throw new ArgumentOutOfRangeException(nameof(logMessage.Level));
+        }
+    }
+
+    private void OnError(IConsumer<Null, byte[]> consumer, Error error)
+    {
+        if (error.IsFatal)
+        {
+            _cts.Cancel();
+        }
+    }
+
+    private Task OnConsumeMessage(ConsumeResult<Null, byte[]> consumeResult,
+        QueueTransportContext queueTransportContext)
+    {
+        var receiveContext = new KafkaReceiveContext(consumeResult, this);
+        IncrementReceiveContextCount();
+        var context = new QueueMessageContext
```

Comment:
You are calling IncrementReceiveContextCount in KafkaTransportPump, but DecrementReceiveContextCount in KafkaReceiveContext. I prefer having the responsibility of increment and decrement be owned by a single class. What do you think about having the constructor of KafkaReceiveContext call the increment instead?

Reply:
done